// @lrnwebcomponents/product-card/lib/hax-element-card-list.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../utils/utils.js", "../../../@material/mwc-switch/mwc-switch.js", "../../../@material/mwc-button/mwc-button.js", "../../../@material/mwc-formfield/mwc-formfield.js", "../../simple-modal/simple-modal.js", "../../hexagon-loader/hexagon-loader.js", "../product-card.js"], function (_exports, _require, _litElement, _utils, _mwcSwitch, _mwcButton, _mwcFormfield, _simpleModal, _hexagonLoader, _productCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXElementCardList = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  class HAXElementCardList extends _litElement.LitElement {
    constructor() {
      super();
      this.showCardList = false;
      this.list = [];
      this.value = {};
      this.cols = 2;
      this._layout = "1-1-1";
      window.SimpleModal.requestAvailability();
    }

    static get tag() {
      return "hax-element-card-list";
    }

    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
          min-height: 100px;
        }
        product-card {
          display: block;
        }
        product-card div[slot="collapse-header"] {
        }
        product-card div[slot="details-collapse-content"] {
          max-height: 125px;
          overflow-y: auto;
        }
        .grid {
          display: grid;
          align-items: stretch;
          grid-template-columns: var(--hax-element-card--cols, repeat(2, 1fr));
          grid-gap: 15px;
          overflow-x: auto;
          padding: 2px;
        }
      `];
    }

    static get properties() {
      return {
        list: {
          type: Array
        },
        enabled: {
          type: Object
        },
        cols: {
          type: Number
        },
        value: {
          type: Object
        },
        hidden: {
          type: Boolean
        },
        showCardList: {
          type: Boolean
        },
        _layout: {
          type: String
        }
      };
    }

    render() {
      return !this.showCardList ? `` : (0, _litElement.html)`
          <hexagon-loader
            item-count="4"
            color="blue"
            ?loading="${this.loading}"
            size="large"
          ></hexagon-loader>
          <h2 ?hidden="${!this.loading}">Loading HAX elements..</h2>
          <div
            class="grid"
            style="--hax-element-card--cols: repeat(${this.cols}, 1fr)"
          >
            ${this.list.map((el, i) => (0, _litElement.html)`
                <product-card
                  .slot="col-${this.__getCol(i)}"
                  ?disabled="${!el.status}"
                  ?has-demo="${el.schema.demoSchema}"
                  heading="${el.schema.gizmo.title}"
                  icon="${el.schema.gizmo.icon}"
                  subheading="${el.schema.gizmo.description}"
                  accent-color="${el.schema.gizmo.color}"
                  data-index="${i}"
                  @product-card-demo-show="${this.toggleShowDemo}"
                  @product-card-demo-hide="${this.toggleShowDemo}"
                >
                  <div class="switch">
                    <mwc-formfield
                      label="${el.status ? `Enabled` : `Disabled`}"
                    >
                      <mwc-switch
                        ?checked="${this.value[tag] === this.value[file]}"
                        @change="${e => this.elementStatusChange(el)}"
                      ></mwc-switch>
                    </mwc-formfield>
                  </div>
                  <div slot="details-collapse-header">Details</div>
                  <div slot="details-collapse-content">
                    <ul>
                      <li>
                        <strong>Tags:</strong> ${el.schema.gizmo.groups.map(group => (0, _litElement.html)`
                              ${group},
                            `)}
                      </li>
                      <li><strong>Tag name:</strong> <code>${el.tag}</code></li>
                      <li>
                        <strong>Developer usage:</strong>
                        <code>import "${el.file}";</code>
                      </li>
                      ${el.schema.gizmo.meta ? (0, _litElement.html)`
                            ${Object.keys(el.schema.gizmo.meta).map(mel => (0, _litElement.html)`
                                <li>
                                  <strong>${this.capFirst(mel)}:</strong>
                                  <span>${el.schema.gizmo.meta[mel]}</span>
                                </li>
                              `)}
                          ` : ``}
                    </ul>
                  </div>
                  <div slot="demo-collapse-header">Demo</div>
                  <div slot="demo-collapse-content">
                    ${el.schema.demoSchema && el.showDemo ? (0, _litElement.html)`
                          ${el.schema.demoSchema.map(demoItem => (0, _litElement.html)`
                              <mwc-button
                                data-tag="${demoItem.tag}"
                                @click="${this._viewDemo}"
                                >Pop up demo</mwc-button
                              >
                              <div class="demo">
                                ${this._haxElementToNode(demoItem)}
                              </div>
                              <code-sample copy-clipboard-button>
                                <template>
                                  ${this._haxElementToNode(demoItem)}
                                </template>
                              </code-sample>
                            `)}
                        ` : ``}
                  </div>
                </product-card>
              `)}
          </div>
        `;
    }

    _viewDemo(e) {
      if (e.target && e.target.nextElementSibling) {
        window.dispatchEvent(new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: "Demo of " + e.target.getAttribute("data-tag"),
            styles: {
              "--simple-modal-width": "75vw",
              "--simple-modal-max-width": "75vw"
            },
            elements: {
              content: e.target.nextElementSibling
            },
            invokedBy: e.target,
            clone: true,
            modal: true
          }
        }));
      }
    }
    /**
     * Effectively event binding to the expanded state
     */


    toggleShowDemo(e) {
      this.list[e.path[0].getAttribute("data-index")].showDemo = e.detail.expanded;
      this.requestUpdate();
    }
    /**
     * Wrap our call so that we can dynamically import code sample since it has a dep tree
     */


    _haxElementToNode(schema) {
      if (schema && schema.tag) {
        new Promise((res, rej) => _require.default(["../../code-sample/code-sample.js"], res, rej));
        return (0, _utils.haxElementToNode)(schema);
      }

      return "";
    }

    capFirst(s) {
      return s.charAt(0).toUpperCase() + s.slice(1);
    }

    elementStatusChange(el, status) {
      if (!status) status = !this.value[el.tag];

      this._updateItem(el.tag, el.file, status); // send up so list can update


      this.dispatchEvent(new CustomEvent("value-changed", {
        detail: this
      }));
      this.requestUpdate();
    }

    __getCol(i) {
      i++;

      while (i > this.cols) {
        i = i - this.cols;
      }

      return i;
    }

    _updateItem(tag, file, status = false) {
      if (!status) {
        delete this.value[tag];
      } else {
        this.value[tag] = file;
      }
    }
    /**
     * LitElement life cycle - property changed
     */


    updated(changedProperties) {
      console.log("updated", this.cols);
      changedProperties.forEach((oldValue, propName) => {
        if (propName == "list" && this.list !== oldValue) {
          this.value = {};
          this.list.forEach(item => this._updateItem(item.tag, item.file, item.status));
        }

        if (propName == "cols") {
          switch (this[propName]) {
            case 3:
              this._layout = "1-1-1";
              break;

            case 4:
              this._layout = "1-1-1-1";
              break;

            case 5:
              this._layout = "1-1-1-1-1";
              break;

            case 6:
              this._layout = "1-1-1-1-1-1";
              break;

            default:
              this._layout = "1-1";
              break;
          }
        }
      });
    }

  }

  _exports.HAXElementCardList = HAXElementCardList;
  customElements.define(HAXElementCardList.tag, HAXElementCardList);
});
