// @lrnwebcomponents/lrndesign-mapmenu/lib/lrndesign-mapmenu-item.js
define(["exports", "../../../lit-element/lit-element.js", "../../../@polymer/iron-icons/iron-icons.js", "../../../@polymer/iron-icon/iron-icon.js"], function (_exports, _litElement, _ironIcons, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignMapmenuItem = void 0;

  function _templateObject3_14d1f5d0b4c611eaada595ac6eca7bce() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          --lrndesign-mapmenu-item-height: 16px;\n        }\n        iron-icon {\n          --iron-icon-height: var(--lrndesign-mapmenu-item-height);\n          display: inline-flex;\n        }\n      "]);

    _templateObject3_14d1f5d0b4c611eaada595ac6eca7bce = function _templateObject3_14d1f5d0b4c611eaada595ac6eca7bce() {
      return data;
    };

    return data;
  }

  function _templateObject2_14d1f5d0b4c611eaada595ac6eca7bce() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <iron-icon icon=\"", "\"></iron-icon>\n          "]);

    _templateObject2_14d1f5d0b4c611eaada595ac6eca7bce = function _templateObject2_14d1f5d0b4c611eaada595ac6eca7bce() {
      return data;
    };

    return data;
  }

  function _templateObject_14d1f5d0b4c611eaada595ac6eca7bce() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n      <span id=\"title\">", "</span>\n    "]);

    _templateObject_14d1f5d0b4c611eaada595ac6eca7bce = function _templateObject_14d1f5d0b4c611eaada595ac6eca7bce() {
      return data;
    };

    return data;
  }

  var LrndesignMapmenuItem =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrndesignMapmenuItem, _LitElement);
    babelHelpers.createClass(LrndesignMapmenuItem, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_14d1f5d0b4c611eaada595ac6eca7bce(), this.icon ? (0, _litElement.html)(_templateObject2_14d1f5d0b4c611eaada595ac6eca7bce(), this.icon) : "", this.title);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject3_14d1f5d0b4c611eaada595ac6eca7bce())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-mapmenu-item";
      }
    }]);

    function LrndesignMapmenuItem() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignMapmenuItem);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignMapmenuItem).call(this));
      _this.icon = "";
      _this.title = "";
      _this.url = "";
      return _this;
    }

    babelHelpers.createClass(LrndesignMapmenuItem, null, [{
      key: "properties",
      get: function get() {
        return {
          icon: {
            type: String
          },
          title: {
            type: String
          },
          url: {
            type: String
          }
        };
      }
    }]);
    return LrndesignMapmenuItem;
  }(_litElement.LitElement);

  _exports.LrndesignMapmenuItem = LrndesignMapmenuItem;
  window.customElements.define(LrndesignMapmenuItem.tag, LrndesignMapmenuItem);
});
