// @lrnwebcomponents/product-card/lib/hax-element-card-list.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../utils/utils.js", "../../../@material/mwc-switch/mwc-switch.js", "../../../@material/mwc-button/mwc-button.js", "../../../@material/mwc-formfield/mwc-formfield.js", "../../simple-modal/simple-modal.js", "../../hexagon-loader/hexagon-loader.js", "../product-card.js"], function (_exports, _require, _litElement, _utils, _mwcSwitch, _mwcButton, _mwcFormfield, _simpleModal, _hexagonLoader, _productCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXElementCardList = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject8_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          min-height: 100px;\n        }\n        product-card {\n          display: block;\n        }\n        product-card div[slot=\"collapse-header\"] {\n        }\n        product-card div[slot=\"details-collapse-content\"] {\n          max-height: 125px;\n          overflow-y: auto;\n        }\n        .grid {\n          display: grid;\n          align-items: stretch;\n          grid-template-columns: var(--hax-element-card--cols, repeat(2, 1fr));\n          grid-gap: 15px;\n          overflow-x: auto;\n          padding: 2px;\n        }\n      "]);

    _templateObject8_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject8_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject7_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                              <mwc-button\n                                data-tag=\"", "\"\n                                @click=\"", "\"\n                                >Pop up demo</mwc-button\n                              >\n                              <div class=\"demo\">\n                                ", "\n                              </div>\n                              <code-sample copy-clipboard-button>\n                                <template>\n                                  ", "\n                                </template>\n                              </code-sample>\n                            "]);

    _templateObject7_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject7_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject6_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          ", "\n                        "]);

    _templateObject6_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject6_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject5_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                <li>\n                                  <strong>", ":</strong>\n                                  <span>", "</span>\n                                </li>\n                              "]);

    _templateObject5_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject5_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject4_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            ", "\n                          "]);

    _templateObject4_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject4_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject3_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                              ", ",\n                            "]);

    _templateObject3_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject3_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject2_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <product-card\n                  .slot=\"col-", "\"\n                  ?disabled=\"", "\"\n                  ?has-demo=\"", "\"\n                  heading=\"", "\"\n                  icon=\"", "\"\n                  subheading=\"", "\"\n                  accent-color=\"", "\"\n                  data-index=\"", "\"\n                  @product-card-demo-show=\"", "\"\n                  @product-card-demo-hide=\"", "\"\n                >\n                  <div class=\"switch\">\n                    <mwc-formfield\n                      label=\"", "\"\n                    >\n                      <mwc-switch\n                        ?checked=\"", "\"\n                        @change=\"", "\"\n                      ></mwc-switch>\n                    </mwc-formfield>\n                  </div>\n                  <div slot=\"details-collapse-header\">Details</div>\n                  <div slot=\"details-collapse-content\">\n                    <ul>\n                      <li>\n                        <strong>Tags:</strong> ", "\n                      </li>\n                      <li><strong>Tag name:</strong> <code>", "</code></li>\n                      <li>\n                        <strong>Developer usage:</strong>\n                        <code>import \"", "\";</code>\n                      </li>\n                      ", "\n                    </ul>\n                  </div>\n                  <div slot=\"demo-collapse-header\">Demo</div>\n                  <div slot=\"demo-collapse-content\">\n                    ", "\n                  </div>\n                </product-card>\n              "]);

    _templateObject2_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject2_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  function _templateObject_04b73c90856511eaa66301b3a4f2e8c5() {
    var data = babelHelpers.taggedTemplateLiteral(["\n          <hexagon-loader\n            item-count=\"4\"\n            color=\"blue\"\n            ?loading=\"", "\"\n            size=\"large\"\n          ></hexagon-loader>\n          <h2 ?hidden=\"", "\">Loading HAX elements..</h2>\n          <div\n            class=\"grid\"\n            style=\"--hax-element-card--cols: repeat(", ", 1fr)\"\n          >\n            ", "\n          </div>\n        "]);

    _templateObject_04b73c90856511eaa66301b3a4f2e8c5 = function _templateObject_04b73c90856511eaa66301b3a4f2e8c5() {
      return data;
    };

    return data;
  }

  var HAXElementCardList =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(HAXElementCardList, _LitElement);

    function HAXElementCardList() {
      var _this;

      babelHelpers.classCallCheck(this, HAXElementCardList);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HAXElementCardList).call(this));
      _this.showCardList = false;
      _this.list = [];
      _this.value = {};
      _this.cols = 2;
      _this._layout = "1-1-1";
      window.SimpleModal.requestAvailability();
      return _this;
    }

    babelHelpers.createClass(HAXElementCardList, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return !this.showCardList ? "" : (0, _litElement.html)(_templateObject_04b73c90856511eaa66301b3a4f2e8c5(), this.loading, !this.loading, this.cols, this.list.map(function (el, i) {
          return (0, _litElement.html)(_templateObject2_04b73c90856511eaa66301b3a4f2e8c5(), _this2.__getCol(i), !el.status, el.schema.demoSchema, el.schema.gizmo.title, el.schema.gizmo.icon, el.schema.gizmo.description, el.schema.gizmo.color, i, _this2.toggleShowDemo, _this2.toggleShowDemo, el.status ? "Enabled" : "Disabled", _this2.value[tag] === _this2.value[file], function (e) {
            return _this2.elementStatusChange(el);
          }, el.schema.gizmo.groups.map(function (group) {
            return (0, _litElement.html)(_templateObject3_04b73c90856511eaa66301b3a4f2e8c5(), group);
          }), el.tag, el.file, el.schema.gizmo.meta ? (0, _litElement.html)(_templateObject4_04b73c90856511eaa66301b3a4f2e8c5(), Object.keys(el.schema.gizmo.meta).map(function (mel) {
            return (0, _litElement.html)(_templateObject5_04b73c90856511eaa66301b3a4f2e8c5(), _this2.capFirst(mel), el.schema.gizmo.meta[mel]);
          })) : "", el.schema.demoSchema && el.showDemo ? (0, _litElement.html)(_templateObject6_04b73c90856511eaa66301b3a4f2e8c5(), el.schema.demoSchema.map(function (demoItem) {
            return (0, _litElement.html)(_templateObject7_04b73c90856511eaa66301b3a4f2e8c5(), demoItem.tag, _this2._viewDemo, _this2._haxElementToNode(demoItem), _this2._haxElementToNode(demoItem));
          })) : "");
        }));
      }
    }, {
      key: "_viewDemo",
      value: function _viewDemo(e) {
        if (e.target && e.target.nextElementSibling) {
          window.dispatchEvent(new CustomEvent("simple-modal-show", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: {
              title: "Demo of " + e.target.getAttribute("data-tag"),
              styles: {
                "--simple-modal-width": "75vw",
                "--simple-modal-max-width": "75vw"
              },
              elements: {
                content: e.target.nextElementSibling
              },
              invokedBy: e.target,
              clone: true,
              modal: true
            }
          }));
        }
      }
      /**
       * Effectively event binding to the expanded state
       */

    }, {
      key: "toggleShowDemo",
      value: function toggleShowDemo(e) {
        this.list[e.path[0].getAttribute("data-index")].showDemo = e.detail.expanded;
        this.requestUpdate();
      }
      /**
       * Wrap our call so that we can dynamically import code sample since it has a dep tree
       */

    }, {
      key: "_haxElementToNode",
      value: function _haxElementToNode(schema) {
        if (schema && schema.tag) {
          new Promise(function (res, rej) {
            return _require.default(["../../code-sample/code-sample.js"], res, rej);
          });
          return (0, _utils.haxElementToNode)(schema);
        }

        return "";
      }
    }, {
      key: "capFirst",
      value: function capFirst(s) {
        return s.charAt(0).toUpperCase() + s.slice(1);
      }
    }, {
      key: "elementStatusChange",
      value: function elementStatusChange(el, status) {
        if (!status) status = !this.value[el.tag];

        this._updateItem(el.tag, el.file, status); // send up so list can update


        this.dispatchEvent(new CustomEvent("value-changed", {
          detail: this
        }));
        this.requestUpdate();
      }
    }, {
      key: "__getCol",
      value: function __getCol(i) {
        i++;

        while (i > this.cols) {
          i = i - this.cols;
        }

        return i;
      }
    }, {
      key: "_updateItem",
      value: function _updateItem(tag, file) {
        var status = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

        if (!status) {
          delete this.value[tag];
        } else {
          this.value[tag] = file;
        }
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        console.log("updated", this.cols);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "list" && _this3.list !== oldValue) {
            _this3.value = {};

            _this3.list.forEach(function (item) {
              return _this3._updateItem(item.tag, item.file, item.status);
            });
          }

          if (propName == "cols") {
            switch (_this3[propName]) {
              case 3:
                _this3._layout = "1-1-1";
                break;

              case 4:
                _this3._layout = "1-1-1-1";
                break;

              case 5:
                _this3._layout = "1-1-1-1-1";
                break;

              case 6:
                _this3._layout = "1-1-1-1-1-1";
                break;

              default:
                _this3._layout = "1-1";
                break;
            }
          }
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "hax-element-card-list";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject8_04b73c90856511eaa66301b3a4f2e8c5())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          list: {
            type: Array
          },
          enabled: {
            type: Object
          },
          cols: {
            type: Number
          },
          value: {
            type: Object
          },
          hidden: {
            type: Boolean
          },
          showCardList: {
            type: Boolean
          },
          _layout: {
            type: String
          }
        };
      }
    }]);
    return HAXElementCardList;
  }(_litElement.LitElement);

  _exports.HAXElementCardList = HAXElementCardList;
  customElements.define(HAXElementCardList.tag, HAXElementCardList);
});
