diff --git a/src/JsonapiParse.php b/src/JsonapiParse.php
index 764e5e0..84d6e85 100644
--- a/src/JsonapiParse.php
+++ b/src/JsonapiParse.php
@@ -99,6 +99,9 @@ class JsonapiParse implements JsonapiParseInterface {
   protected function flattenIncluded($resource) {
     if (isset($this->included[$resource['type']][$resource['id']])) {
       $object = $this->resolveAttributes($this->included[$resource['type']][$resource['id']]);
+      if (isset($resource['meta'])) {
+        $object['meta'] = $resource['meta'];
+      }
     }
     else {
       $object = $resource;
